metric_learn/_util.py

```diff
@@ -12,39 +14,264 @@ def vector_norm(X):
+      else:
+        make_error_input(201, input, context)
+    elif input.ndim == 3:  # we should check_num_features which is not checked
+                           #  after
+      pass
+    else:
+      if preprocessor is not None:
+        make_error_input(420, input, context)
+      else:
+        make_error_input(200, input, context)
+
+    input = check_array(input, allow_nd=True, ensure_2d=False,
+                        **args_for_sk_checks)
+    if ensure_min_features > 0:
+      n_features = input.shape[2]
+      if n_features < ensure_min_features:
+        raise ValueError("Found array with {} feature(s) (shape={}) while"
+                         " a minimum of {} is required{}."
+                         .format(n_features, input.shape,
+                                 ensure_min_features, context))
+    #  normally we don't need to check_t too because t should'nt be able to
+    # be modified by any preprocessor
+    if input.ndim != 3:  # we have to ensure this because check_array above
+      # does not
+      if preprocessor_has_been_applied:
+        make_error_input(211, input, context)
+      else:
+        make_error_input(201, input, context)
+    check_t(input, t, context)
+
+  return input if y is None else (input, y)
+
+
+def make_error_input(code, input, context):
+  code_str = {'expected_input': {'1': '2D array of formed points',
+                                 '2': '3D array of formed tuples',
+                                 '3': ('1D array of indicators or 2D array of '
+                                       'formed points'),
+                                 '4': ('2D array of indicators or 3D array '
+                                       'of formed tuples')},
+              'additional_context': {'0': '',
+                                     '2': ' when using a preprocessor',
+                                     '1': (' after the preprocessor has been '
+                                           'applied')},
+              'possible_preprocessor': {'0': '',
+                                        '1': ' and/or use a preprocessor'
+                                        }}
+  code_list = str(code)
+  err_args = dict(expected_input=code_str['expected_input'][code_list[0]],
+                  additional_context=code_str['additional_context']
+                  [code_list[1]],
+                  possible_preprocessor=code_str['possible_preprocessor']
+                  [code_list[2]],
+                  input=input, context=context, found_size=input.ndim)
+  err_msg = ('{expected_input} expected'
+             '{context}{additional_context}. Found {found_size}D array '
+             'instead:\ninput={input}. Reshape your data'
+             '{possible_preprocessor}.\n')
+  raise ValueError(err_msg.format(**err_args))
+
+
+def preprocess_tuples(tuples, preprocessor):
+  print("Preprocessing tuples...")
+  tuples = np.column_stack([preprocessor(tuples[:, i])[:, np.newaxis] for

```

Comment:
I don't think you need to do the `np.newaxis` stuff here, as `np.column_stack` will accept 1d inputs.

Reply:
Actually the output of preprocessor(1d input of shape (n_constraints,)) is a 2d input (of shape (n_constraints, n_features). This is the way I found to concatenate t=`tuple_size` 2d arrays into a 3d array of shape (n_constraints, tuple_size, n_features), but there may be a better way